compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/PostponedArgumentsAnalyzer.kt

```diff
@@ -170,23 +170,29 @@ class PostponedArgumentsAnalyzer(
     ) {
         val (returnArguments, inferenceSession) = results
 
-        returnArguments.forEach { c.addSubsystemFromExpression(it) }
         val checkerSink: CheckerSink = CheckerSinkImpl(candidate)
         val builder = c.getBuilder()
 
         val lastExpression = lambda.atom.body?.statements?.lastOrNull() as? FirExpression
         var hasExpressionInReturnArguments = false
-        // No constraint for return expressions of lambda if it has Unit return type.
-        val lambdaReturnType = lambda.returnType.let(substitute).takeUnless { it.isUnitOrFlexibleUnit }
+        val lambdaReturnType = lambda.returnType.let(substitute)
         returnArguments.forEach {
+            val haveSubsystem = c.addSubsystemFromExpression(it)
             if (it !is FirExpression) return@forEach
+            // If the lambda returns Unit, the last expression is not returned and should not be constrained.
+            // TODO? questionable moment inherited from FE1.0 (the `haveSubsystem` case):
```

Comment:
Please, add a YT issue and link it here just to track it.
Current K1 behavior looks weird indeed, so may be it's worth trying redesigning it.

Reply:
https://youtrack.jetbrains.com/issue/KT-55837